src/algebra/group/defs.lean

```diff
@@ -592,39 +613,48 @@ zpow_coe_nat a n
 theorem zpow_neg_succ_of_nat (a : G) (n : ℕ) : a ^ -[1+n] = (a ^ (n + 1))⁻¹ :=
 by { rw ← zpow_coe_nat, exact div_inv_monoid.zpow_neg' n a }
 
-end
+@[to_additive] lemma div_eq_mul_inv (a b : G) : a / b = a * b⁻¹ := div_inv_monoid.div_eq_mul_inv _ _
 
-@[to_additive]
-lemma div_eq_mul_inv {G : Type u} [div_inv_monoid G] :
-  ∀ a b : G, a / b = a * b⁻¹ :=
-div_inv_monoid.div_eq_mul_inv
+alias div_eq_mul_inv ← division_def
 
-section
--- ensure that we don't go via these typeclasses to find `has_inv` on groups and groups with zero
-set_option extends_priority 50
+end div_inv_monoid
 
-/-- Auxiliary typeclass for types with an involutive `has_inv`. -/
-@[ancestor has_inv]
-class has_involutive_inv (G : Type*) extends has_inv G :=
-(inv_inv : ∀ x : G, x⁻¹⁻¹ = x)
+/-- A `subtraction_monoid` is a `sub_neg_monoid` with involutive negation and such that
+`-(a + b) = -b + -a`. -/
+@[protect_proj, ancestor sub_neg_monoid has_involutive_neg]
+class subtraction_monoid (G : Type u) extends sub_neg_monoid G, has_involutive_neg G :=
+(neg_add_rev (a b : G) : -(a + b) = -b + -a)
+(neg_eq_of_add (a b : G) : a + b = 0 → -a = b)
 
-/-- Auxiliary typeclass for types with an involutive `has_neg`. -/
-@[ancestor has_neg]
-class has_involutive_neg (A : Type*) extends has_neg A :=
-(neg_neg : ∀ x : A, - -x = x)
+/-- A `division_monoid` is a `div_inv_monoid` with involutive inversion and such that
+`(a * b)⁻¹ = b⁻¹ * a⁻¹`.
 
-attribute [to_additive] has_involutive_inv
+This is the immediate common ancestor of `group` and `group_with_zero`. -/
+@[protect_proj, ancestor div_inv_monoid has_involutive_inv, to_additive subtraction_monoid]
+class division_monoid (G : Type u) extends div_inv_monoid G, has_involutive_inv G :=
+(mul_inv_rev (a b : G) : (a * b)⁻¹ = b⁻¹ * a⁻¹)
+(inv_eq_of_mul (a b : G) : a * b = 1 → a⁻¹ = b)
```

Comment:
I wonder if it makes things easier to provide instances if you make this
```suggestion
(inv_eq_of_mul (a b : G) : a * b = 1 → b * a = 1 → a⁻¹ = b)
```
Is this requirement equivalent or stronger?

Reply:
The only point would be to prove `division_monoid (mul_opposite \a)`, right? It's easy enough (see branch `stronger_div_inv_monoid`) that I wouldn't change the definition for that.

It's not clear to me whether this is equivalent or weaker. It's definitely not stronger however.

Reply:
The other point would be to make it clear that this class is not just about right inverses.

Reply:
If you have a short proof that this is equivalent, I'm happy to switch. But I suspect it's weaker.

Reply:
Thanks for the example! but you didn't show that you could define `⁻¹` consistently with the other axioms.

And my `division_monoid` **does** give two-sided inverses. I haven't proved it in this PR because I wanted to only provide the barebone definition, but `a * b = 1 → b⁻¹ = a` (because `⁻¹` is involutive). Further, Eric's definition is (a priori) weaker, so it goes against the spirit of this typeclass.

Reply:
> Further, Eric's definition is (a priori) weaker, so it goes against the spirit of this typeclass.

It's only against the spirit of the typeclass if it's actually weaker. Typeclasses that express their fields as something weak and derive something strong are often convenient.

Reply:
> you didn't show that you could define `⁻¹` consistently with the other axioms.

Good point. If we ignore `mul_inv_rev` then we could define `f⁻¹` to be the two-sided inverse if `f` is bijective, otherwise `f` itself: this is involutive and satisfies only the two-sided `inv_eq_of_mul`. If we also want to satisfy `mul_inv_rev` I can't think of a good example.

Reply:
> And my `division_monoid` **does** give two-sided inverses. I haven't proved it

Indeed, one proof is:
```lean
lemma inv_one {G} [division_monoid G] : (1 : G)⁻¹ = 1 :=
begin
  apply inv_eq_of_mul,
  rw mul_one,
end

lemma inv_eq_of_mul' {G} [division_monoid G] (a b : G) : b * a = 1 → a⁻¹ = b :=
begin
  intro h,
  replace h := congr_arg has_inv.inv h,
  rw [mul_inv_rev, inv_one] at h,
  have :=  congr_arg has_inv.inv  (inv_eq_of_mul _ _ h),
  simpa only [inv_inv] using this,
end
```

Reply:
This probably doesn't need to block this PR, as the weaker field is only a nice-to-have: but it would be good to understand whether my definition is truly weaker.

If we keep yours, can you add a comment that despite `inv_eq_of_mul` looking asymmetric, the other axioms ensure the symmetric version is also true?